### PR TITLE
feat: Remove defer

### PR DESCRIPTION
**Pull request by markbeierl**
# Description

Removes explicit certificate relations and defer events

# Checklist:

- [X] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [X] I have performed a self-review of my own code
I have made corresponding changes to the documentation
- [X] I have added tests that validate the behaviour of the software
- [X] I validated that new and existing unit tests pass locally with my changes
Any dependent changes have been merged and published in downstream modules
I have bumped the version of the library

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -110,31 +110,23 @@ def __init__(self, *args):
...
         )
-        self.framework.observe(
-            self._certificates.on.certificate_available, self._on_certificate_available
-        )
         self.framework.observe(
             self._certificates.on.certificate_expiring, self._on_certificate_expiring
         )
 
-    def _configure_nrf(self, event: EventBase) -> None:
+    def _configure_nrf(self, event: EventBase) -> None:  # noqa C901
```
Comment (ghislainbourgeois) [inline src/charm.py:122]: I would suggest not inhibiting `C901` and finding a refactoring that will solve it instead.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 2]
```diff
@@ -145,32 +137,32 @@ def _configure_nrf(self, event: EventBase) -> None:
...
         if not _get_pod_ip():
             self.unit.status = WaitingStatus("Waiting for pod IP address to be available")
-            event.defer()
             return
-        if not self._certificate_is_stored():
+        if not self._private_key_is_stored():
+            self._generate_private_key()
+        if not self._csr_is_stored():
+            self._request_new_certificate()
+        certificate_changed = False
```
Comment (gatici) [inline src/charm.py:151]: We can remove line 151.  The reason:
if provider_certificate exists, `certificate_changed` variable will be defined in line 154.
If provider certificate does not exist, return will exit the method and line 162 does not run unless line 153 returns True.

File: src/charm.py [hunk 3]
```diff
@@ -226,6 +191,31 @@ def _on_database_relation_broken(self, event: EventBase) -> None:
...
+    def _update_certificate(self, provider_certificate) -> bool:
+        """Compares the provided certificate to what is stored.
+
+        Returns True if the certificate was updated
+        """
+        existing_certificate = (
+            self._get_stored_certificate() if self._certificate_is_stored() else ""
+        )
+
+        if not existing_certificate == provider_certificate:
```
Comment (gatici) [inline src/charm.py:214]:
```suggestion
        if existing_certificate != provider_certificate:
```

File: src/charm.py (hunk 2 above)
Comment (markbeierl) [reply src/charm.py:151]: That just feels wrong to me to not define a variable in the scope that it is going to be used.  Anyone else have thoughts on this?

File: src/charm.py (hunk 1 above)
Comment (markbeierl) [reply src/charm.py:122]: I appreciate not adding `#noqa C901`, but I really don't know what would make a _configure() function have fewer if/then statements.  Do I group related ones and make a function call for them?

File: src/charm.py (hunk 2 above)
Comment (patriciareinoso) [reply src/charm.py:151]: I also feel uncomfortable if we do not define the variable.

File: src/charm.py (hunk 1 above)
Comment (patriciareinoso) [reply src/charm.py:122]: I can see 2 groups of actions : precondition checks and actual configuration actions

Comment (markbeierl) [reply src/charm.py:122]: ok, but essentially does that really remove complexity?  We would need a method to return true/false if the preconditions are met, and stop if not.  Then we need another one for configuration that returns true/false for the restart.  Essentially we are adding complexity with additional logic branches, no?

File: src/charm.py [hunk 4]
```diff
@@ -145,32 +137,32 @@ def _configure_nrf(self, event: EventBase) -> None:
             return
         if not self._get_database_uri():
             self.unit.status = WaitingStatus("Waiting for database URI")
-            event.defer()
             return
         if not self._container.exists(path=BASE_CONFIG_PATH):
```
Comment (patriciareinoso) [inline src/charm.py:141]: it just came to my attention that we verify that `/etc/nrf/` is mounted but we do not verify `/support/TLS`
should we check both ?

Comment (markbeierl) [reply src/charm.py:141]: good point, yes, I would think so

Comment (markbeierl) [reply src/charm.py]: done

File: src/charm.py (hunk 1 above)
Comment (markbeierl) [reply src/charm.py]: Broke out preconditions

File: src/charm.py (hunk 2 above)
Comment (markbeierl) [reply src/charm.py]: Avoided the issue with a refactor

**Review by gatici — APPROVED** [review]
LGTM. The branch needs to be rebased.

**Review by ghislainbourgeois — APPROVED** [review]

**Pull request merged**